lib/handler/file.c

```diff
@@ -110,6 +127,57 @@ static void do_proceed(h2o_generator_t *_self, h2o_req_t *req)
         do_close(&self->super, req);
 }
 
+static void do_multirange_proceed(h2o_generator_t *_self, h2o_req_t *req) {
+    struct st_h2o_sendfile_generator_t *self = (void *)_self;
+    size_t rlen, *range_cur = self->ranged.range_infos + 2 * self->ranged.finished_range, used_buf = 0;
+    ssize_t rret, vecarrsize;
+    h2o_iovec_t vec, *vecarr;
+    int is_finished;
+
+    if (self->ranged.current_range == self->ranged.finished_range) {
+        size_t range_end = *range_cur + *(range_cur + 1) - 1;
+        if (H2O_LIKELY(self->ranged.current_range != 0))
+            used_buf = sprintf(self->buf, "\r\n--%s\r\nContent-Type: %s\r\nContent-Range: bytes %zd-%zd/%zd\r\n\r\n",
+                               self->ranged.boundary.base, self->ranged.mimetype.base, *range_cur, range_end,
+                               self->ranged.filesize);
+        else
+            used_buf = sprintf(self->buf, "--%s\r\nContent-Type: %s\r\nContent-Range: bytes %zd-%zd/%zd\r\n\r\n",
+                               self->ranged.boundary.base, self->ranged.mimetype.base, *range_cur, range_end,
+                               self->ranged.filesize);            
+        self->ranged.current_range ++;
+        lseek(self->fd, *range_cur, SEEK_SET);
+    }
+    range_cur ++;
+    rlen = *range_cur;
+    if (rlen + used_buf > MAX_BUF_SIZE)
+        rlen = MAX_BUF_SIZE - used_buf;
+    while ((rret = read(self->fd, self->buf + used_buf, rlen)) == -1 && errno == EINTR)
+        ;
+    if (rret == -1) {
+        req->http1_is_persistent = 0;
+        h2o_send(req, NULL, 0, 1);
+        do_close(&self->super, req);
+        return;
+    }
+    *range_cur -= rret;
+    if (*range_cur == 0) self->ranged.finished_range ++;
+
+    vec.base = self->buf;
+    vec.len = rret + used_buf;
+    if (self->ranged.finished_range == self->ranged.range_count) {
+        vecarr = (void *)h2o_mem_alloc_pool(&req->pool, 2 * sizeof(h2o_iovec_t));
+        *vecarr = vec;
+        *(vecarr + 1) = self->ranged.end_marker_buf;
+        vecarrsize = 2;
+        is_finished = 1;
+    } else {
+        vecarr = &vec;
+        vecarrsize = 1;
+        is_finished = 0;
+    }
```

Comment:
How about declaring a `vec` as `h2o_iovec_t [2]`, and always using it (with `vecarrsize` set to `1` or `2`) instead of using `vec`, `vecarr` and calling `h2o_mem_alloc_pool`?